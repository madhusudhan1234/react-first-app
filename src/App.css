body {
  background: #ffc600;
}
.page {
  margin: 20px;
  }
  .interactions {
  text-align: center;
  }
  .table {
  margin: 20px 0;
  }
  .table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 10px 0;
  background: #F9F9F9;
  border-bottom: 1px solid #F1F1F1;
  justify-content: space-between;
  }
  .table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
  }
  .table-row {
    display: flex;
    line-height: 24px;
    white-space: nowrap;
    padding: 10px;
    background: #ffffff;
    border-bottom: 1px solid #F1F1F1;
    }
    .table-header > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
    }
    .table-row > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
    }
    .button-inline {
    border-width: 0;
    background: transparent;
    color: inherit;
    text-align: inherit;
    -webkit-font-smoothing: inherit;
    padding: 0;
    font-size: inherit;
    cursor: pointer;
    }
    button[type="Button"] {
      background: #ffffff;
    }
    .button-active {
    border-radius: 0;
    border-bottom: 1px solid #38BB6C;
    }