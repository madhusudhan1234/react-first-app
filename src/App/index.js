import {
  SubmitButton,
  SaveButton,
  CancelButton
} from '../Buttons';