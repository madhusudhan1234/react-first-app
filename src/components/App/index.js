import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  DEFAULT_BASE,
  PATH_SEARCH,
  PARAM_PAGE,
  PARAM_HPP,
} from '../constants';
import {
  SubmitButton,
  SaveButton,
  CancelButton
} from '../Buttons';