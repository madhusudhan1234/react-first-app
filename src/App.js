import React, { Component } from 'react';
import fetch from 'isomorphic-fetch';
import propTypes from 'prop-types';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_HPP = '100';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}`;

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
    };

    this.needsToSearchTopStories = this.needToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }

  needToSearchTopStories(searchTerm) {
    return  !this.state.results[searchTerm];
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }
    event.preventDefault();
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits = results && results[searchKey]
      ? results[searchKey].hits
      : [];

    const updatedHits = [
      ...oldHits,
      ...hits
    ];
    
    this.setState({ 
      results: { 
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
     });
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(e => this.setState({ error: e }));
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];
    const isNotId = item => item.objectID !== id;

    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    })
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }
  
  render() {
    const { 
      searchTerm, 
      results, 
      searchKey,
      error
    } = this.state;
    
    const page = (
      results && 
      results[searchKey] && 
      results[searchKey].page
    ) || 0;

    const list = (
      results && 
      results[searchKey] && 
      results[searchKey].hits
    ) || [];

    if (error) {
      return <p>Something went wrong.</p>;
    }

    return (
      <div className="page">
        <div className="interactions">
            <Search 
              value={searchTerm}
              onChange={this.onSearchChange}
              onSubmit={this.onSearchSubmit}
            >
              Search
            </Search>
          { error 
            ? <div className="interactions">
              <p> Something went wrong. </p>
            </div>
            : <Table 
              list={list}
              pattern={searchTerm}
              onDismiss={this.onDismiss}
            />
          }
          <div className="interactions">
            <Button onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
              More
            </Button>  
          </div>
        </div>
      </div>
    );
  }
}

const Search = ({
    value,
    onChange,
    onSubmit,
    children 
  }) => 
      <form onSubmit={onSubmit}>
        {children} <input
          type="text"
          value={value}
          onChange={onChange}
        />
        <button type="submit">
          {children}
        </button>
      </form>

const Button = ({ onClick, className, children }) => 
  <button
    onClick={onClick}
    className={className}
    type="button"
  >  
    {children}
  </button>


Button.propTypes = {
  onClick: propTypes.func.isRequired,
  className: propTypes.string,
  children: propTypes.node.isRequired,
};  

Button.defaultProps = {
  className: '',
};

const Table = ({ list, onDismiss }) => {
  const largeColumn = {
    width: '40%',
  }

  const midColumn = {
    width: '30%',
  }

  const smallColumn = {
    width: '10%',
  }

  return (
    <div className='table'>
    {list.map(item => 
      <div key={item.objectID} className='table-row'>
        <span style={largeColumn}>
          <a href={item.url}>{item.title}</a>
        </span>
        <span style={midColumn}>{item.author}</span>
        <span style={smallColumn}>{item.num_comments}</span>
        <span style={smallColumn }>{item.points}</span>
        <span style={smallColumn}>
          <Button onClick={() => onDismiss(item.objectID)} className='button-inline'>
            Dismiss
          </Button>  
        </span>
      </div>
    )}
  </div>
  )
}

Table.propTypes = {
  list: propTypes.arrayOf(
    propTypes.shape({
      objectID: propTypes.string.isRequired,
      author: propTypes.string,
      url: propTypes.string,
      num_comments: propTypes.number,
      points: propTypes.number,
    })
  ).isRequired,
  onDismiss: propTypes.func.isRequired,
};

export default App;
export { Button, Search, Table };
