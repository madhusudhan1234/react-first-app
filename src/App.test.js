import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App, { Search } from './App';

describe('App', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    //ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Search', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDom.render(<Search>Search</Search>, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Search>Search</Search>
    );
    let tree = compoenent.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

