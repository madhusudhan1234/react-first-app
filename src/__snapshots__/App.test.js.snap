// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      Search
       
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
    <div
      className="table"
    >
      <div
        className="table-header"
      >
        <span
          style={
            Object {
              "width": "40%",
            }
          }
        >
          Title  
          <button
            className="button-inline"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </span>
        <span
          style={
            Object {
              "width": "30%",
            }
          }
        >
          Author  
          <button
            className="button-inline"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </span>
        <span
          style={
            Object {
              "width": "10%",
            }
          }
        >
          Comments  
          <button
            className="button-inline"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </span>
        <span
          style={
            Object {
              "width": "10%",
            }
          }
        >
          Points  
          <button
            className="button-inline"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-sort fa-w-10 "
              data-icon="sort"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </span>
        <span
          style={
            Object {
              "width": "10%",
            }
          }
        >
          Archive
        </span>
      </div>
    </div>
    <div
      className="interactions"
    >
      <div>
        Loading...
      </div>
    </div>
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give Me More
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  Search
   
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      Title  
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Author  
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Comments  
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Points  
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sort fa-w-10 "
          data-icon="sort"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-archive fa-w-16 "
          data-icon="archive"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M32 448c0 17.7 14.3 32 32 32h384c17.7 0 32-14.3 32-32V160H32v288zm160-212c0-6.6 5.4-12 12-12h104c6.6 0 12 5.4 12 12v8c0 6.6-5.4 12-12 12H204c-6.6 0-12-5.4-12-12v-8zM480 32H32C14.3 32 0 46.3 0 64v48c0 8.8 7.2 16 16 16h480c8.8 0 16-7.2 16-16V64c0-17.7-14.3-32-32-32z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-archive fa-w-16 "
          data-icon="archive"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M32 448c0 17.7 14.3 32 32 32h384c17.7 0 32-14.3 32-32V160H32v288zm160-212c0-6.6 5.4-12 12-12h104c6.6 0 12 5.4 12 12v8c0 6.6-5.4 12-12 12H204c-6.6 0-12-5.4-12-12v-8zM480 32H32C14.3 32 0 46.3 0 64v48c0 8.8 7.2 16 16 16h480c8.8 0 16-7.2 16-16V64c0-17.7-14.3-32-32-32z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </span>
  </div>
</div>
`;
